pkg/file/file.go

```diff
@@ -25,6 +28,13 @@ func SaveFileWithBackup(path string, data []byte, backupDir string) error {
 		return errors.Errorf("%s is directory", path)
 	}
 
+	if _, ok := fileLocks[path]; !ok {
```

Comment:
The fileLocks itself is not thread safe, so I think the datarace is still exists. eg.

```bash
# assume the fileLocks is empty as initial state.
# thread 1 start
if _, ok := fileLocks["/path/a"]; !ok { # the ok is false because fileLocks is empty
# thread 1 hang after check !ok but before fileLocks[path] = &sync.Mutex{}

# thread 2 start
if _, ok := fileLocks["/path/a"]; !ok { # the ok is false because fileLocks is empty and thread 1 doesn't assign it yet
    fileLocks["/path/a"] = &sync.Mutex{} # assign a new mutex and success
}

fileLocks["/path/a"].Lock()
# thread 2 enter critical area

# thread 1 start
fileLocks["/path/a"] = &sync.Mutex{} # assign a new mutex and success
fileLocks["/path/a"].Lock() # this will success because thread 1 reset fileLocks["/path/a"]
# thread 1 enter critical area
```

In this case, both threads enter into the same critical area.

Reply:
Thanks for your suggestion, sorry to forget that the map needs to be locked in concurrent access.